spec/requests/comments_n_plus_one_spec.rb

```diff
@@ -0,0 +1,114 @@
+require "spec_helper"
+
+describe "n+1 queries in the CommentsController" do
+  include LoginMacros
+
+  shared_examples "a hierarchical view with constant queries" do
+    context "when displaying multiple comments", n_plus_one: true do
+      context "when the comments are flat" do
+        populate { |n| create_list(:comment, n, commentable: commentable) }
+
+        warmup { subject.call }
+
+        it "produces a constant number of queries when logged out" do
+          expect do
+            subject.call
+          end.to perform_constant_number_of_queries
+        end
+
+        it "produces a constant number of queries when logged in" do
+          fake_login
+
+          expect do
+            subject.call
+          end.to perform_constant_number_of_queries
+        end
+      end
+
+      context "when the comments are nested" do
+        populate do |n|
+          n.times.inject(commentable) do |parent|
+            create(:comment, commentable: parent)
+          end
+        end
+
+        warmup { subject.call }
+
+        it "produces a constant number of queries when logged out" do
+          expect do
+            subject.call
+          end.to perform_constant_number_of_queries
+        end
+
+        it "produces a constant number of queries when logged in" do
+          fake_login
+
+          expect do
+            subject.call
+          end.to perform_constant_number_of_queries
+        end
+      end
+    end
+  end
+
+  describe "#show_comments" do
+    subject do
+      proc do
+        get show_comments_comments_path(work_id: work, format: :js), xhr: true
+      end
+    end
+
+    let!(:work) { create(:work) }
+    let!(:commentable) { work.first_chapter }
+
+    it_behaves_like "a hierarchical view with constant queries"
+  end
+
+  describe "#show" do
+    subject do
+      proc do
+        get comment_path(comment)
+      end
+    end
+
+    let!(:comment) { create(:comment) }
+    let!(:commentable) { comment }
+
+    it_behaves_like "a hierarchical view with constant queries"
+  end
+
+  describe "#index" do
+    subject do
+      proc do
+        get work_comments_path(work, show_comments: true)
+      end
+    end
+
+    let!(:work) { create(:work) }
+    let!(:commentable) { work.first_chapter }
+
+    it_behaves_like "a hierarchical view with constant queries"
+  end
+
+  describe "#unreviewed" do
+    let!(:work) { create(:work, moderated_commenting_enabled: true) }
+
+    context "when displaying multiple comments", n_plus_one: true do
+      before { fake_login_known_user(work.users.first) }
+
+      populate do |n|
+        User.current_user = nil # prevent bugs with unreviewed check
```

Comment:
Which bugs does this comment refer to?

Reply:
The unreviewed check in comments relies on `User.current_user`, which is set when making a request through the `ApplicationController`, but not cleared after the request is complete. This means that any checks that rely on `User.current_user` that run after a request by a logged-in user will behave as if it's the logged-in user performing those actions, even if those actions are a call to `FactoryBot.create`.

In this case, the sequence of events is:
1) 2 comments are created on the moderated work. `User.current_user` is nil, so they're marked as unreviewed.
2) The owner of the work views those two unreviewed comments, causing `User.current_user` to become set.
3) The 2 comments are deleted.
4) 3 new comments are created on the moderated work. Because `User.current_user` is set to the owner of the work, these comments count as creator comments, even though the pseuds on the comments are not creator pseuds. As a result, none of the comments are unreviewed. So you end up comparing the display of 2 comments vs. the display of 0 comments, instead of 2 vs. 3 as it should be. (This is, incidentally, why I added the check involving `current_scale`.)

The issue is widespread -- that's why most of the calls to `FactoryBot.create(:work)` in the feature tests are preceded by a call to `step %{I start a new session}`, because if `User.current_user` was still set, they'd fail with an error about the creator not allowing co-creator requests.

At some point I'd like to convert the `set_current_user` hook to an `around_action`, so that it can clear `User.current_user` after the request is over. But until then, it's sometimes going to be necessary to either log out, or manually set `User.current_user` to nil.

Reply:
Makes sense, thanks! Yeah, we'll need to deal with `User.current_user` before we [start using request specs more heavily](https://stackoverflow.com/questions/40851705/controller-specs-vs-request-specs).

Can you change the comment to "log out to prevent commenting as work creator"?